src/sage/rings/lazy_series.py

```diff
@@ -5390,6 +5390,75 @@ def derivative(self, *args):
                                                    P.is_sparse())
         return P.element_class(P, coeff_stream)
 
+    def adams_operator(self, p):
+        """
+        Return the image of ``self`` under the Adams operator of index ``p``.
+
+        This raises all variables to the power ``p``, both the power
+        series variables and the variables inside the coefficient ring.
+
+        INPUT:
+
+        - ``p`` -- a positive integer
+
+        EXAMPLES:
+
+        With no variables in the base ring::
+
+            sage: A = LazyPowerSeriesRing(QQ,'t')
+            sage: f = A([1,2,3,4])
+            sage: f.adams_operator(2)
+            1 + 2*t^2 + 3*t^4 + 4*t^6
+
+        With variables in the base ring::
+
+            sage: q = polygen(QQ,'q')
+            sage: A = LazyPowerSeriesRing(q.parent(),'t')
+            sage: f = A([0,1+q,2,3+q**2])
+            sage: f.adams_operator(2)
+            ((q^2+1)*t^2) + 2*t^4 + ((q^4+3)*t^6)
+
+        In the multivariate case::
+
+            sage: A = LazyPowerSeriesRing(ZZ,'t,u')
+            sage: f = A({(1,2):4,(2,3):6})
+            sage: f.adams_operator(3)
+            4*t^3*u^6 + 6*t^6*u^9
+
+        TESTS::
+
+            sage: A = LazyPowerSeriesRing(QQ,'t')
+            sage: f = A([1,2,3,4])
+            sage: f.adams_operator(1)
+            1 + 2*t + 3*t^2 + 4*t^3
+            sage: f.adams_operator(-1)
+            Traceback (most recent call last):
+            ...
+            ValueError: p must be a positive integer
+        """
+        if p <= 0:
+            raise ValueError("p must be a positive integer")
+
+        if p == 1:
+            return self
+
+        try:
+            D = {v: v**p for v in self.base_ring().gens()}
+        except AttributeError:
+            with_vars = False
+        else:
+            with_vars = True
+
+        scaled = self(*[g**p for g in self.parent().gens()])
+
+        if not with_vars:
+            return scaled
+
+        def func(cf):
+            return cf.subs(D)
+
+        return scaled.map_coefficients(func)
```

Comment:
Maybe turning around the logic is easier?
```
        scaled = self(*[g**p for g in self.parent().gens()])
        try:
            vars = self.base_ring().gens()
        except AttributeError:
            return scaled
        D = {v: v**p for v in vars}
        return scaled.map_coefficients(lambda c: c.subs(D))
```

Reply:
Let me again mention that basically all reasonable base rings have a `gens()` attribute. So the `try-except` block is effectively worthless (although perhaps it is okay to have it for the rare (if it even exists) commutative ring that does not implement `gens()`). The good news is that `subs` also works fine on things like `ZZ` and `QQ`. It introduces an inefficiency, but I doubt it is a big one.